x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/registry/ClearInferenceEndpointCacheAction.java

```diff
@@ -0,0 +1,239 @@
+        );
+
+        static {
+            PARSER.declareLong(ConstructingObjectParser.constructorArg(), VERSION_FIELD);
+        }
+
+        public static InvalidateCacheMetadata fromXContent(XContentParser parser) {
+            return PARSER.apply(parser, null);
+        }
+
+        public static InvalidateCacheMetadata fromMetadata(ProjectMetadata projectMetadata) {
+            InvalidateCacheMetadata metadata = projectMetadata.custom(NAME);
+            return metadata == null ? EMPTY : metadata;
+        }
+
+        private final long version;
+
+        private InvalidateCacheMetadata(long version) {
+            this.version = version;
+        }
+
+        public InvalidateCacheMetadata(StreamInput in) throws IOException {
+            this(in.readVLong());
+        }
+
+        public InvalidateCacheMetadata bumpVersion() {
+            return new InvalidateCacheMetadata(version < Long.MAX_VALUE ? version + 1 : 1L);
+        }
+
+        @Override
+        public EnumSet<Metadata.XContentContext> context() {
+            return Metadata.ALL_CONTEXTS;
+        }
+
+        @Override
+        public TransportVersion getMinimalSupportedVersion() {
+            return ML_INFERENCE_ENDPOINT_CACHE;
+        }
+
+        @Override
+        public String getWriteableName() {
+            return NAME;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            out.writeVLong(version);
+        }
+
+        @Override
+        public Iterator<? extends ToXContent> toXContentChunked(ToXContent.Params ignored) {
+            return Iterators.single(((builder, params) -> builder.field(VERSION_FIELD.getPreferredName(), version)));
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hashCode(version);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (other == this) return true;
+            return other instanceof InvalidateCacheMetadata that && that.version == this.version;
+        }
+    }
+
+    private static class RefreshCacheMetadataVersionTask extends AckedBatchedClusterStateUpdateTask {
+        private final ProjectId projectId;
+
+        private RefreshCacheMetadataVersionTask(ProjectId projectId, ActionListener<AcknowledgedResponse> listener) {
+            super(TimeValue.THIRTY_SECONDS, listener);
+            this.projectId = projectId;
+        }
+    }
+
+    private static class CacheMetadataUpdateTaskExecutor extends SimpleBatchedAckListenerTaskExecutor<RefreshCacheMetadataVersionTask> {
+        @Override
+        public Tuple<ClusterState, ClusterStateAckListener> executeTask(RefreshCacheMetadataVersionTask task, ClusterState clusterState) {
+            var projectMetadata = clusterState.metadata().getProject(task.projectId);
+            var currentMetadata = InvalidateCacheMetadata.fromMetadata(projectMetadata);
+            var updatedMetadata = currentMetadata.bumpVersion();
```

Comment:
Why do we need to keep track of a version and bump it? Is it because we wouldn't change the cluster state otherwise?